### PR TITLE
Make IP address optional when reporting transactions

### PR DESCRIPTION
**Pull request by ugexe**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,14 @@
 CHANGELOG
 =========
 
+3.1.0
+------------------
+
+* Updated `MaxMind\MinFraud\ReportTransaction` to make the `ip_address`
+  parameter optional. Now the `tag` and at least one of the following
+  parameters must be supplied: `ip_address`, `maxmind_id`, `minfraud_id`,
+  `transaction_id`.
```
Comment (oschwald) [inline CHANGELOG.md:10]:
```suggestion
  parameters must be supplied: `ipAddress`, `maxmindId`, `minfraudId`,
  `transactionId`.
```

File: README.md [hunk 2]
```diff
@@ -282,8 +282,9 @@ All externally visible exceptions are in the `\MaxMind\Exception` namespace.
 The possible exceptions are:
 
 * `InvalidInputException` - This will be thrown when the `->report()` method is
-  called with invalid input data or when the required `ip_address` or `tag`
-  fields are missing.
+  called with invalid input data or when the required fields are missing. The
+  required fields are `tag` and one or more of the following: `ip_address`,
+  `maxmind_id`, `minfraud_id`, or `transaction_id`.
```
Comment (oschwald) [inline README.md:287]:
```suggestion
  called with invalid input data or when the required arguments are null. The
  required arguments are `tag` and one or more of the following: `ipAddress`,
  `maxmindId`, `minfraudId`, or `transactionId`.
```

This one is my fault. I should have updated this when I switched to named arguments as the preferred method.

File: src/MinFraud/ReportTransaction.php [hunk 3]
```diff
@@ -46,9 +46,11 @@ public function __construct(
      * @param array  $values         An array of transaction parameters. The keys are the same
      *                               as the JSON keys. You may use either this or the named
      *                               arguments, but not both.
-     * @param string $ipAddress      Required. The IP address of the customer placing the
+     * @param string $ipAddress      Optional. The IP address of the customer placing the
      *                               order. This should be passed as a string like
-     *                               "44.55.66.77" or "2001:db8::2:1".
+     *                               "44.55.66.77" or "2001:db8::2:1". This field is not
```
Comment (oschwald) [inline src/MinFraud/ReportTransaction.php:51]: Also on me, but these "field" uses probably should all be "argument". It might be better for us to just do that in a separate PR though as I suspect it is endemic.

File: src/MinFraud/ReportTransaction.php [hunk 4]
```diff
@@ -59,22 +61,25 @@ public function __construct(
...
      *                               additional details to help us understand context are
      *                               helpful.
      * @param string $transactionId  Optional. The transaction ID you originally passed to
-     *                               minFraud. This field is not required, but you are
-     *                               encouraged to provide it or the transaction's
-     *                               maxmind_id or minfraud_id.
+     *                               minFraud. This field is not required if you provide at
+     *                               least one of the transaction's ipAddress, maxmindId, or
+     *                               minfraudId. You are encouraged to provide it or the
+     *                               transaction's maxmind_id or minfraud_id.
```
Comment (oschwald) [inline src/MinFraud/ReportTransaction.php:82]:
```suggestion
     *                               transaction's maxmindId or minfraudId.
```

I suppose we could put backticks around all of these, but that is preexisting.

File: src/MinFraud/ReportTransaction.php [hunk 5]
```diff
@@ -59,22 +61,25 @@ public function __construct(
...
+     *                               ipAddress, minfraudId, or transactionId. You are
+     *                               encouraged to provide it, if possible.
      * @param string $minfraudId     Optional. A UUID that identifies a minFraud Score,
      *                               minFraud Insights, or minFraud Factors request. This
      *                               ID is returned at /id in the response. This field is
-     *                               not required, but you are encouraged to provide it if
-     *                               the request was made to one of these services.
+     *                               not required if you provide at least one of the transaction's
+     *                               ipAddress, maxmindId, or transactionId. You are encouraged to
+     *                               provide it the request was made to one of these services.
```
Comment (oschwald) [inline src/MinFraud/ReportTransaction.php:72]:
```suggestion
     *                               provide it if the request was made to one of these services.
```

File: src/MinFraud/ReportTransaction.php [hunk 6]
```diff
@@ -116,18 +121,27 @@ public function report(
             $this->verifyEmpty($values);
         }
 
+        if ($ipAddress === null
+            && $minfraudId === null
```
Comment (oschwald) [inline src/MinFraud/ReportTransaction.php:125]: Given that we are not validating that this is a valid UUID, maybe w should at least check that it is not the mepty string.

**Review by oschwald — CHANGES_REQUESTED** [review]

File: src/MinFraud/ReportTransaction.php (hunk 6 above)
Comment (ugexe) [reply src/MinFraud/ReportTransaction.php:125]: `$minfraudId` is validated as a UUID a little later in the validation process
https://github.com/maxmind/minfraud-api-php/blob/a0f49df251a8d82edf8399bb3311300344f4ca1e/src/MinFraud/ReportTransaction.php#L153-L162

That being said, `$maxmindId` is also validated to ensure it is 8 characters so I don't think we need the check for `$maxmindId === ''`. It might also make sense to put this "one of these fields" check after the aforementioned individual field validation has occurred.

**Review by oschwald — APPROVED** [review]
